cli/admin/src/main/java/io/pravega/cli/admin/readerGroup/ParseReaderGroupStreamCommand.java

```diff
@@ -0,0 +1,173 @@
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+
+public class ParseReaderGroupStreamCommand extends AdminCommand {
+
+    private final static int HEADER = WireCommands.TYPE_SIZE;
+    private final static int LENGTH = WireCommands.TYPE_PLUS_LENGTH_SIZE - WireCommands.TYPE_SIZE;
+    private final static int TYPE = WireCommandType.EVENT.getCode();
+    private final static int REQUEST_TIMEOUT_SECONDS = 10;
+
+    private final GrpcAuthHelper authHelper;
+
+    public ParseReaderGroupStreamCommand(CommandArgs args) {
+        super(args);
+
+        authHelper = new GrpcAuthHelper(true, "secret", 600);
+    }
+
+    @Override
+    public void execute() throws Exception {
+        ensureArgCount(4);
+        final String scope = getArg(0);
+        final String readerGroup = getArg(1);
+        final String segmentStoreHost = getArg(2);
+        final String fileName = getArg(3);
+        String stream = NameUtils.getStreamForReaderGroup(readerGroup);
+
+        @Cleanup
+        CuratorFramework zkClient = createZKClient();
+        @Cleanup
+        SegmentHelper segmentHelper = instantiateSegmentHelper(zkClient);
+        String tmpfilename = "tmp/RG_" + readerGroup + "_" + new Random().nextInt(1000);
+        readRGSegmentToFile(segmentHelper, segmentStoreHost, scope, stream, tmpfilename, fileName);
+        output("The readerGroup stream has been successfully written into %s", fileName);
+    }
+
+    /**
+     * Reads the contents of the segment starting from the given offset and writes into the provided file.
+     *
+     * @param segmentHelper       A {@link SegmentHelper} instance to read the segment.
+     * @param segmentStoreHost    Address of the segment-store to read from.
+     * @param scope               The name of the scope.
+     * @param stream              The name of the stream.
+     * @param fileName            A name of the file to which the data will be written.
+     * @throws IOException if the file create/write fails.
+     * @throws Exception if the request fails.
+     */
+    private void readRGSegmentToFile(SegmentHelper segmentHelper, String segmentStoreHost, String scope, String stream,
+                                     String tmpfilename, String fileName) throws IOException, Exception {
+
+        File outputfile = FileHelper.createFileAndDirectory(fileName);
+
+        @Cleanup
+        BufferedWriter writer = new BufferedWriter(new FileWriter(outputfile));
+
+        Controller controller = instantiateController();
+        CompletableFuture<StreamSegments> streamSegments = controller.getCurrentSegments(scope, stream);
+        StreamSegments ss = streamSegments.get(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        int segmentid = (ss.getNumberOfSegments() > 0) ? (ss.getNumberOfSegments() - 1) : 0;
```

Comment:
Are we sure that the segments of a reader group are always identified by increasing numbers without exception? I suppose that it may be the behavior of state synchronizer, but please check that the assumption you are making here holds, even for reader groups with many (rolled) segments.

Also, could this expression be simplified as `int segmentId = ss.getNumberOfSegments() - 1`? A reader group stream will have always at least 1 segment, right?

Reply:
modified this